pkgs/jnigen/lib/src/bindings/dart_generator.dart

```diff
@@ -769,38 +778,42 @@ class _TypeGenerator extends TypeVisitor<String> {
 
   @override
   String visitDeclaredType(DeclaredType node) {
-    final nullable = includeNullability && node.isNullable ? '?' : '';
-    if (node.classDecl.binaryName == 'java.lang.Object' ||
-        node.classDecl.binaryName == 'java.lang.String') {
-      return '$_jni.${node.classDecl.finalName}$nullable';
+    if (node.classDecl.isObject) {
+      // The class is not generated, fall back to `JObject`.
+      return super.visitDeclaredType(node);
     }
+    final nullable =
+        includeNullability && node.isNullable && isTopTypeNullable ? '?' : '';
 
-    // All type parameters of this type
+    // All type parameters of this type.
     final allTypeParams = node.classDecl.allTypeParams
         .map((typeParam) => typeParam.name)
         .toList();
     // The ones that are declared.
-    final typeGenerator = _TypeGenerator(
-      resolver,
-      forInterfaceImplementation: forInterfaceImplementation,
-      typeErasure: forInterfaceImplementation,
-      includeNullability: true,
-      arrayType: false,
-    );
-    final definedTypeParams = node.params.accept(typeGenerator).toList();
+    final definedTypeParams = <String>[];
+    for (var i = 0; i < node.params.length; ++i) {
+      final param = node.params[i];
+      final index = allTypeParams.length - node.params.length + i;
```

Comment:
This allTypeParams stuff feels like it should be factored out into a little helper class or something. You have little bits of logic for dealing with allTypeParams indexes spread throughout this function. Better to consolidate that logic if possible.

Reply:
Good suggestion, rewriting the logic made the code much better!